compiler/optimizer/OMRRecognizedCallTransformer.cpp

```diff
@@ -0,0 +1,61 @@
+/*******************************************************************************
+* Copyright (c) 2017, 2017 IBM Corp. and others
+*
+* This program and the accompanying materials are made available under
+* the terms of the Eclipse Public License 2.0 which accompanies this
+* distribution and is available at http://eclipse.org/legal/epl-2.0
+* or the Apache License, Version 2.0 which accompanies this distribution
+* and is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* This Source Code may also be made available under the following Secondary
+* Licenses when the conditions for such availability set forth in the
+* Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+* version 2 with the GNU Classpath Exception [1] and GNU General Public
+* License, version 2 with the OpenJDK Assembly Exception [2].
+*
+* [1] https://www.gnu.org/software/classpath/license.html
+* [2] http://openjdk.java.net/legal/assembly-exception.html
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*******************************************************************************/
+
+#include "optimizer/RecognizedCallTransformer.hpp"
+
+#include "il/Node.hpp"
+#include "il/Node_inlines.hpp"
+#include "il/TreeTop.hpp"
+#include "il/TreeTop_inlines.hpp"
+#include "codegen/CodeGenerator.hpp" // for CodeGenerator
+
+TR::Optimization* OMR::RecognizedCallTransformer::create(TR::OptimizationManager *manager)
+   {
+   return new (manager->allocator()) TR::RecognizedCallTransformer(manager);
+   }
+
+int32_t OMR::RecognizedCallTransformer::perform()
+   {
+   TR_BitVector visited;
+   visited.init(0, comp()->trMemory(), heapAlloc, growable);

```

Comment:
This BitVector should not use heapAlloc - it is sufficient to use stack alloc so the memory is freed at the end of the optimization instead of at the end of the compilation. The correct declaration would be something like:
`TR_BitVector visited(comp()->trMemory()->currentStackRegion());`
There is a separate facility that already exists in the compilation to give you a special kind of BitVector that is recycled by the compilation for the purposes of doing this kind of tracking. The advantage of this special type is that the compilation manages a pool of them which saves constantly having to allocate them. To use these you want to look in the OMR project at the compiler/infra/Checklist.hpp header. In there you will find `NodeChecklist` and `BlockChecklist`. You would just want to do:
`TR::NodeChecklist visited(comp());`
That checklist has appropriate `add` `contains` and `remove` methods which will make things easy.

If you could have a look and rework to use the `TR::NodeChecklist` that would be great since that is the direction we want to move to for checking for repeated visits.

Reply:
Wonderful, the `TR::NodeChecklist` approach is much cleaner than the `TR_BitVector` one.  Updated.